[UI-side compositing] Have pinch zoom constrain zoom origin in commitTransientZoom

nmoucht opened the pull request:
#### d1ac2b0c842b877514d0109437565162ebf7272b
<pre>
[UI-side compositing] Have pinch zoom constrain zoom origin in commitTransientZoom
<a href="https://bugs.webkit.org/show_bug.cgi?id=252345">https://bugs.webkit.org/show_bug.cgi?id=252345</a>
rdar://105515035

Reviewed by Simon Fraser.

Mirror TiledCoreAnimationDrawingArea code by constraining the zoom origin to prevent
view jumping after commiting a zoom in certain scenarios. We do this by querying
the root frame scrolling node about certain state to pass to the ScrollableArea
static function constrainScrollPositionForOverhang(). We also pass this constrained
origin to the WebProcess. We also add several functions to ScrollingTree and
RemoteScrollingCoordinatorProxy to retreive root node state while holding the tree
lock. It was also necessary to change ScrollingTree::mainFrameScrollPosition to use
the root node&apos;s scroll position rather than mainFrameScrollPosition since that
member does not appear to be updated with UI-side compositing on.

* Source/WebCore/page/scrolling/ScrollingTreeFrameScrollingNode.h:
* Source/WebKit/UIProcess/RemoteLayerTree/mac/RemoteLayerTreeDrawingAreaProxyMac.mm:
(WebKit::RemoteLayerTreeDrawingAreaProxyMac::commitTransientZoom):
* Source/WebKit/WebProcess/WebPage/mac/RemoteLayerTreeDrawingAreaMac.mm:
(WebKit::RemoteLayerTreeDrawingAreaMac::commitTransientZoom):

Canonical link: <a href="https://commits.webkit.org/261226@main">https://commits.webkit.org/261226@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/de931d21e5b7593bbd9baa9529ff164fa2b6c10d

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/110812 "2 style errors") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/19898 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/43356 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/87/builds/2343 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/119709 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/114774 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/21295 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/85/builds/10999 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/86/builds/2025 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/116562 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/78/builds/15866 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/99026 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/103212 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/9/builds/97826 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/73/builds/30692 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/44259 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/12504 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/32028 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/34/builds/86142 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/13064 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/84/builds/9008 "Passed tests") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/18438 "Built successfully") | [  ~~🧪 mac-wk2-stress~~](https://ews-build.webkit.org/#/builders/62/builds/51649 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/15009 "Built successfully") | | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/75/builds/4255 "Built successfully and passed tests") | | | | 
<!--EWS-Status-Bubble-End-->

smfr commented on Source/WebKit/UIProcess/RemoteLayerTree/mac/RemoteLayerTreeDrawingAreaProxyMac.mm:
```diff
@@ -206,20 +208,38 @@ explicit RemoteLayerTreeDisplayLinkClient(WebPageProxyIdentifier pageID)
-
+    auto rootNode = downcast<ScrollingTreeFrameScrollingNode>(m_webPageProxy.scrollingCoordinatorProxy()->rootNode());
+    auto scrollPosition = rootNode->currentScrollPosition();
+    auto visibleContentRect = rootNode->layoutViewport();
```
These two things are not the same. The layout viewport is larger than the visible content rect when zoomed in, so I don't think this is the right rectangle. I would expect some visible UI issues with this code.

nmoucht replied:
Yeah I was a bit confused by this but when I tried out the TiledCoreAnimationDrawingArea code I saw that the `frameView.visibleContentRectIncludingScrollbars` was always returning the viewport size, irrespective of scale. This is because `FrameView::visibleContentScaleFactor()` always returns 1 in this case. It is possible that this is a bug, but in both cases (UI-side compositing off and on) I am seeing the snapping behavior being properly prevented. At the very least it is improperly named.

smfr commented on Source/WebKit/UIProcess/RemoteLayerTree/mac/RemoteLayerTreeDrawingAreaProxyMac.mm:
```diff
@@ -206,20 +208,38 @@ explicit RemoteLayerTreeDisplayLinkClient(WebPageProxyIdentifier pageID)
 {
-    LOG_WITH_STREAM(ViewGestures, stream << "RemoteLayerTreeDrawingAreaProxyMac::commitTransientZoom - scale " << scale << " origin " << origin);
-
+    auto rootNode = downcast<ScrollingTreeFrameScrollingNode>(m_webPageProxy.scrollingCoordinatorProxy()->rootNode());
```
Prefer `dynamicDowncast` with a null check to avoid security bugs triggered by untrusted webcontent data.

nmoucht replied:
Fixed.

smfr commented on Source/WebKit/UIProcess/RemoteLayerTree/mac/RemoteLayerTreeDrawingAreaProxyMac.mm:
```diff
@@ -206,20 +208,38 @@ explicit RemoteLayerTreeDisplayLinkClient(WebPageProxyIdentifier pageID)
+    auto scrollPosition = rootNode->currentScrollPosition();
+    auto visibleContentRect = rootNode->layoutViewport();
+    visibleContentRect.setX(scrollPosition.x());
+    visibleContentRect.setY(scrollPosition.y());
```
Better to factor the code above into a helper called `computeVisibleContentRect()` or similar

nmoucht replied:
Fixed.

smfr commented on Source/WebKit/UIProcess/RemoteLayerTree/mac/RemoteLayerTreeDrawingAreaProxyMac.mm:
```diff
@@ -206,20 +208,38 @@ explicit RemoteLayerTreeDisplayLinkClient(WebPageProxyIdentifier pageID)
+    constrainedOrigin.moveBy(-origin);
+
+    IntSize scaledTotalContentsSize = roundedIntSize(rootNode->totalContentsSize());
+    scaledTotalContentsSize.scale(scale / rootNode->frameScaleFactor());
```
Is `rootNode->frameScaleFactor()` the last scale that we committed to the web process? Is that what we want?

nmoucht replied:
Yes we want to scale this by the difference in scale, based on how TiledCoreAnimationDrawingArea uses `scale / m_webPage.totalScaleFactor()`.

nmoucht committed:
[UI-side compositing] Have pinch zoom constrain zoom origin in commitTransientZoom
https://bugs.webkit.org/show_bug.cgi?id=252345
rdar://105515035

Reviewed by Simon Fraser.

Mirror TiledCoreAnimationDrawingArea code by constraining the zoom origin to prevent
view jumping after commiting a zoom in certain scenarios. We do this by querying
the root frame scrolling node about certain state to pass to the ScrollableArea
static function constrainScrollPositionForOverhang(). We also pass this constrained
origin to the WebProcess. We also add several functions to ScrollingTree and
RemoteScrollingCoordinatorProxy to retreive root node state while holding the tree
lock. It was also necessary to change ScrollingTree::mainFrameScrollPosition to use
the root node's scroll position rather than mainFrameScrollPosition since that
member does not appear to be updated with UI-side compositing on.

* Source/WebCore/page/scrolling/ScrollingTreeFrameScrollingNode.h:
* Source/WebKit/UIProcess/RemoteLayerTree/mac/RemoteLayerTreeDrawingAreaProxyMac.mm:
(WebKit::RemoteLayerTreeDrawingAreaProxyMac::commitTransientZoom):
* Source/WebKit/WebProcess/WebPage/mac/RemoteLayerTreeDrawingAreaMac.mm:
(WebKit::RemoteLayerTreeDrawingAreaMac::commitTransientZoom):

Canonical link: https://commits.webkit.org/261226@main
